docs/cvl/expr.md

```diff
@@ -550,3 +550,116 @@ As with direct storage access in general, direct storage havoc is experimental a
 * user-defined types such as structs, or arrays/mappings of such types
 * enums
 ```
+
+## Built-in Functions
+
+### Hashing
+
+Many contracts employ hashing in order to create `mapping` keys from more complex data structures.
+A native hashing function in Solidity is `keccak256`.
+We expose `keccak256` in CVL, in order to be able to create such mapping keys directly in CVL.
+The alternative to using the CVL built-in for `keccak256` is to use pre-existing
+functionality from the contract, or creating a dedicated harness function,
+which is usually undesirable from both usability and Prover performance perspectives.
+
+Currently, only the `keccak256` hash is supported in CVL as a built-in.
+
+#### Example
+
+Given the following Solidity snippet:
+```solidity
+contract HashingExample {
+  struct SignedMessage {
+    address sender;
+    uint256 nonce;
+    bytes signature;
+  }
+
+  mapping (bytes32 => uint256) messageToValue;
+
+  function hashingScheme1(SignedMessage memory s) public pure returns (bytes32) {
+    return keccak256(abi.encode(s.sender, s.nonce));
+  }
+
+  function hashingScheme2(SignedMessage memory s) public pure returns (bytes32) {
+    return keccak256(s.signature);
+  }
+
+  function hashingScheme3(SignedMessage memory s) public pure returns (bytes32) {
+    return keccak256(abi.encode(s.sender, s.nonce, keccak256(s.signature)));
+  }
+
+  function hashingScheme4(SignedMessage memory s) public pure returns (bytes32) {
+    return keccak256(abi.encode(s.sender, s.nonce, s.signature));
+  }
+}
+```
+
+The hashing schemes described by `hashingScheme1`, `hashingScheme2`, and `hashingScheme3` can be replicated in CVL as follows:
+```
+function hashingScheme1CVL(HashingExample.SignedMessage s) returns bytes32 {
+  return keccak256(s.sender, s.nonce);
+}
+
+function hashingScheme2CVL(HashingExample.SignedMessage s) returns bytes32 {
+  return keccak256(s.signature);
+}
+
+function hashingScheme3CVL(HashingExample.SignedMessage s) returns bytes32 {
+  return keccak256(s.sender, s.nonce, keccak256(s.signature));
+}
+```
+
+The scheme implemented in `hashingScheme4` is not supported at the moment, as it combines a `bytes` type with primitives.
+In general, the `keccak256` built-in function supports two kinds of inputs:
```

Comment:
I'm not sure what "in general" means here, it suggests that there are some exceptions to this rule, which there are not.

Reply:
great, then I read the code correctly.